src/handle.rs

```diff
@@ -101,31 +105,35 @@ impl<'scope> Mutator<'scope> {
     /// [`Atomic`]: struct.Atomic.html
     pub fn pin<F, R>(&self, f: F) -> R
     where
-        F: FnOnce(&Scope) -> R,
+        F: for<'scope> FnOnce(Scope<'scope>) -> R,
```

Comment:
Take a look at signatures of [`scope` in Rayon](https://docs.rs/rayon/0.8.2/rayon/fn.scope.html) and [`scope` in Crossbeam](https://docs.rs/crossbeam/0.3.0/crossbeam/fn.scope.html). It probably doesn't matter, but let's consider writing our signature in a similar manner:
```rust
pub fn pin<'scope, F, R>(&self, f: F) -> R
where
    F: FnOnce(Scope<'scope>) -> R
```

Also, I don't know how to feel about the change from `&'scope Scope` to `Scope<'scope>`. :)

On one hand, `Scope<'scope>` seems sensible: the motivation for parametrizing over the lifetime is the fact that we now have the `global: &'scope Global` field inside `Scope`. But on the other hand, we still have this `bag: *mut Bag`, which is technically just `bag: &'scope mut Bag` in disguise. So this raw pointer is a missed opportunity for using the lifetime.

Why don't we simplify a bit and define `Scope` as:
```rust
struct Scope {
    handle: *const Handle,
}

// or like this:

struct Scope<'scope> {
    handle: &'scope Handle,
}
```

and then reach for `bag` and `global` through the handle?

By the way, note that Rayon's `Scope<'scope>` has `'scope` only because of the `PhantomData` field (see [here](https://docs.rs/rayon-core/1.2.1/src/rayon_core/scope/mod.rs.html#18)), but is otherwise using raw pointers. And Crossbeam's scoped-thread `Scope` doesn't really need `'scope` - it could be implemented without the lifetime as well.

What do you think?

Reply:
- Type signature of `pin()`: After changing the type as you described, I couldn't satisfy the compiler for the following line: https://github.com/jeehoonkang/crossbeam-epoch/blob/handle/src/default.rs#L27 .  What do you think?

- Having `handle` instead of `global` and `bag` in `Scope`: I like the idea.  It should have been implemented in that way from the beginning :)

  But interestingly, the current branch may be better performance-wise, because its access pattern is more shallow than your proposal.  In your proposal, `global` should be accessed **after** `handle` is resolved.  On the other hand, the current branch doesn't have such a dependency.

  I think, however, its hypothetical performance benefit is far less than the benefit of the clarity of your proposal.  I'll change the PR now.

- Type definition of `Scope`: I prefer `handle: &'scope Handle`, because, as you said, it better uses the lifetime logic.  (I actually used `Option<&'scope Handle>` in order to support `fn unprotected()`.  I checked that it is optimized to have the same size with `usize`.